### PR TITLE
[ci] compile with -quick & validate after vio2vo

### PR DESCRIPTION
**Pull request by gares**
I'd love CI maintainers to help me here, since I'm coding by copy/pasting.

What I think I've done:
- one extra build with -quick plus a call to -vio2vo
- one extra validate on top of the new build

What may be worth adding as an extra "chain":
1. regular build of vo files (could we reuse the artifacts of another job?)
2. clean the .vo, keep the .aux
3. build -quick
4. -vio2vo
5. validate

The interest would be that in the current job we only delay proofs that are in an empty section (if they don't have Proof using). If we have .aux files around we delay all proofs, hence we test more.

### CODE REVIEW

File: .gitlab-ci.yml
Comment (SkySkimmer) [inline .gitlab-ci.yml]: I think this overrides the script from the build-template.

File: .gitlab-ci.yml
Comment (SkySkimmer) [inline .gitlab-ci.yml]: Putting this in the test stage makes it impossible to use it in the validate job.

File: .gitlab-ci.yml
Comment (gares) [reply .gitlab-ci.yml]: is there a way to *append* ?

File: .gitlab-ci.yml
Comment (gares) [reply .gitlab-ci.yml]: ok thanks I'll fix this.

File: .gitlab-ci.yml
Comment (gares) [reply .gitlab-ci.yml]: I need the regular build + this extra command.

**Comment by gares** [comment]
CC: #8642

File: .gitlab-ci.yml
Comment (SkySkimmer) [reply .gitlab-ci.yml]: I don't think so.

**Comment by SkySkimmer** [comment]
It may be better to just have a combined job in the test stage with a custom script
~~~bash
./configure -local -warn-error yes -native-compiler yes
COQUSERFLAGS=-quick make -j $NJOBS
bin/coqc -boot -schedule-vio2vo 4 $(find . -name '*.vio')
~~~

File: .gitlab-ci.yml
Comment (SkySkimmer) [inline .gitlab-ci.yml]: Why coqide?

File: .gitlab-ci.yml
Comment (gares) [reply .gitlab-ci.yml]: I found `after_script:`

Comment (SkySkimmer) [reply .gitlab-ci.yml]: Using after_script breaks the bot.

File: .gitlab-ci.yml
Comment (gares) [reply .gitlab-ci.yml]: removed

**Comment by SkySkimmer** [comment]
Does building with `-quick` work properly? It seems fragile considering that the make rules expect it to produce vo but it will actually produce vio. For instance I'm worried about the following `make validate` deciding to rebuild some vo: imagine b depends on a, and schedule-vio2vo finishes b.vo first, then `make validate` will see b as being outdated.

**Comment by gares** [comment]
> then `make validate` will see b as being outdated.

You are right, indeed I did run coqchk by hand on my machine, I'll follow your suggestion.

**Comment by gares** [comment]
wait, validate does not run `make validate`, so there is no rebuilding

**Comment by SkySkimmer** [comment]
>wait, validate does not run make validate, so there is no rebuilding

Right, I forgot that.
I guess the install rules don't depend on the stuff they install so that's not an issue.

**Review by SkySkimmer — DISMISSED** [review]
Using after_script causes problems for the bot.

**Comment by Zimmi48** [comment]
@SkySkimmer 
>Using `after_script` causes problems for the bot.

This is not the case anymore. However, `after_script` should *not* be used to append to `script` because this is not its semantics (it will run even if `script` fails).

**Comment by gares** [comment]
I think the PR is ok now @SkySkimmer

File: .gitlab-ci.yml
Comment (ejgallego) [inline .gitlab-ci.yml]: All this custom logic doesn't belong into the CI scripts [think for example we would like to call from Azure]

**Review by ejgallego — CHANGES_REQUESTED** [review]

File: .gitlab-ci.yml
Comment (SkySkimmer) [reply .gitlab-ci.yml]: We want to run -quick from azure?

File: .gitlab-ci.yml
Comment (SkySkimmer) [inline .gitlab-ci.yml]: I wouldn't bother with those echos, they just make the script harder to read. The ones in the build template are just there on inertia from travis.

File: .gitlab-ci.yml
Comment (ejgallego) [reply .gitlab-ci.yml]: Do we support `-quick` on Windows?

Comment (gares) [reply .gitlab-ci.yml]: > All this custom logic doesn't belong into the CI scripts

Do we support validate on Azure? (this line is very close to the one of validate)

**Comment by gares** [comment]
> All this custom logic doesn't belong into the CI scripts [think for example we would like to call from Azure]

I'm happy to fix the PR but I need to understand what you mean by "custom logic".

File: .gitlab-ci.yml
Comment (gares) [reply .gitlab-ci.yml]: Which dependency path? I'm sorry but I don't understand. What would be the problem on Azure in the first place?

Comment (ejgallego) [reply .gitlab-ci.yml]: Ideally the first target would be produced by the makefile just by depending on `VOFILES:.vo:=.vio`, then, the second step should also be produced from a make rule that knows how to do `.vio -> .vo`.

**Comment by gares** [comment]
@ejgallego I've pushed the logic to the Makefile and tested locally:
```
$ QUICK=1 make -j2 stdlib
COQC -quick theories/Init/Notations.vo
...
COQC -quick plugins/setoid_ring/Rings_R.v
COQC -schedule-vio2vo theories/**vio plugins/**vio

$ make validate
COQCHK <theories & plugins>
```
Instead of exposing a target to build all .vio and then another one to obtain the .vo I opted for `stdlib` meaning I want all vos, and if you are in QUICK mode I'll do the vio2vo step.

@coq/ci-maintainers I get an issue on gitlab I can't quite understand, help is welcome (where does the . come from?)

**Comment by SkySkimmer** [comment]
>@coq/ci-maintainers I get an issue on gitlab I can't quite understand, help is welcome (where does the . come from?)

When COQ_EXTRA_CONF is unset configure gets an empty argument.
Use `COQ_EXTRA_CONF: "-native-compiler no"` or some such thing.

**Comment by SkySkimmer** [comment]
We already use QUICK in makefile for doc stuff, I don't think it's a good idea to reuse the variable.

**Review by ejgallego — APPROVED** [review]
I am not an expert on Makefiles but this is OK to me. I wonder why the `touch` is needed tho.

**Comment by gares** [comment]
> We already use QUICK in makefile for doc stuff, I don't think it's a good idea to reuse the variable.

Isn't the same semantics? "Do as fast as you can something with less quality guarantee?"

> When COQ_EXTRA_CONF is unset configure gets an empty argument.
> Use `COQ_EXTRA_CONF: "-native-compiler no"` or some such thing.

a HUGE thanks

**Comment by gares** [comment]
> I wonder why the `touch` is needed tho.

it avoids that subsequent runs of make think .vos are outdated.

it should be probably something that should be done by vio2vo, that generates .vo files in random order (or better optimal scheduling order if you have aux file around). I'll add a comment and open an issue about that.

**Comment by SkySkimmer** [comment]
>Isn't the same semantics? "Do as fast as you can something with less quality guarantee?"

Is it actually quicker?

**Comment by gares** [comment]
It depends on the project and the number of cores (and the Proof using annotations/aux files present).
Here on CI it is faster, but I guess it is mostly because of native-compiler no.
The figures on the odd order theorem were decent: 10x on a 12 core (page 10 https://hal.inria.fr/hal-01135919/ )

**Review by ppedrot — COMMENTED** [review]
Not qualified to review.

**Comment by gares** [comment]
Ci failure is spurious, so I guess we could consider this PR for merge. It lets us detect regressions in -quick and -schedule-vio2vo/-vio2vo and, more importantly, that `Module Include` stays out of the opaque proofs business, and one can eventually start working on coq/ceps#40 (I guess after SProp and native integers are merged, to avoid conflicts)

**Comment by gares** [comment]
I was preparing a separate PR fixing #9334 but then I've decided to push the commit here

File: stm/vio_checking.ml
Comment (SkySkimmer) [inline stm/vio_checking.ml]: If there's an error do we end up writing some random successful vos? If so then shouldn't we still update their times?

Comment (gares) [reply stm/vio_checking.ml]: That is the faster review ever ;-)
This code does exactly what the makefile line was doing.

I frankly don't know how to improve things. Removing some .vo files that were successfully built seems to bring nothing. But I don't get how setting the times of some .vo could help. I mean if you then manage to obtain the remaining vos, theirs date will differ anyway, confusing make.

So unless you have some other argument, I'd stay like that.

Comment (SkySkimmer) [reply stm/vio_checking.ml]: >But I don't get how setting the times of some .vo could help

If a prefix succeeds it might make sense (eg everything in Init/ succeeds but somewhere else fails, then `make states` can tell nothing needs to be done)
Not a huge gain though

**Comment by SkySkimmer** [comment]
Gitlab windows job failed to download sed for some reason, I'll retry it.

**Comment by gares** [comment]
ping (CI failures seem totally unrelated)

File: .gitlab-ci.yml
Comment (SkySkimmer) [inline .gitlab-ci.yml]: Do we want -native-compiler yes here?

Comment (gares) [reply .gitlab-ci.yml]: I don't think so, it does not compile opaque proofs hence we would just slow down the build imo

Comment (SkySkimmer) [reply .gitlab-ci.yml]: OK, let's merge. If someone wants native compiler they can always make a subsequent PR.

**Comment by SkySkimmer** [comment]
I caught a segfault in a CI run: https://gitlab.com/SkySkimmer/coq/-/jobs/147407854
~~~
Makefile.build:821: recipe for target 'theories/Reals/Cauchy_prod.vo' failed
make[1]: *** [theories/Reals/Cauchy_prod.vo] Segmentation fault (core dumped)
~~~

**Pull request merged**
